### PR TITLE
Homework 34

### PR DESCRIPTION
**Pull request by vremere-a**

### CODE REVIEW

**Review by KostiantynVorobiov — APPROVED** [review]
OK!

File: src/main/java/com/internet/shop/model/User.java [hunk 1]
```diff
@@ -32,7 +33,7 @@ public User(Long id, String name, String surname, String email, String login, St
     }
 
     public User(Long id, String name, String surname,
-                 String email, String login, String password, Set<Role> roles) {
+                String email, String login, String password, Set<Role> roles) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/model/User.java:33]: image
Let's modify User constructors like this

File: src/main/resources/init_db.sql [hunk 2]
```diff
@@ -93,5 +103,17 @@ CREATE TABLE `internet_shop`.`users_roles` (
...
-ALTER TABLE `internet_shop`.`users`
-    ADD COLUMN `deleted` TINYINT NOT NULL DEFAULT 0 AFTER `password`;
+CREATE TABLE `internet_shop`.`users` (
+                         `user_id` bigint NOT NULL AUTO_INCREMENT,
+                         `user_name` varchar(255) NOT NULL,
+                         `user_surname` varchar(255) NOT NULL,
+                         `email` varchar(255) NOT NULL,
+                         `login` varchar(255) NOT NULL,
+                         `password` varchar(255) NOT NULL,
+                         `salt` blob,
```
Comment (dmytro-hryhoruk) [inline src/main/resources/init_db.sql]: let's add constrain not null to this field too

File: src/main/resources/init_db.sql [hunk 3]
```diff
@@ -21,20 +21,30 @@ CREATE TABLE `internet_shop`.`orders` (
 
 
 CREATE TABLE `internet_shop`.`orders_products` (
```
Comment (dmytro-hryhoruk) [inline src/main/resources/init_db.sql]: reformat this file(remove these tabs). And delete comments

File: src/main/java/com/internet/shop/model/User.java (hunk 1 above)
Comment (vremere-a) [reply src/main/java/com/internet/shop/model/User.java:33]: done

File: src/main/resources/init_db.sql (hunk 2 above)
Comment (vremere-a) [reply src/main/resources/init_db.sql]: done

File: src/main/resources/init_db.sql (hunk 3 above)
Comment (vremere-a) [reply src/main/resources/init_db.sql]: done

File: src/main/resources/init_db.sql [hunk 4]
```diff
@@ -93,5 +90,17 @@ CREATE TABLE `internet_shop`.`users_roles` (
...
+                         `user_name` varchar(255) NOT NULL,
+                         `user_surname` varchar(255) NOT NULL,
+                         `email` varchar(255) NOT NULL,
+                         `login` varchar(255) NOT NULL,
+                         `password` varchar(255) NOT NULL,
+                         `salt` blob NOT NULL,
+                         `deleted` tinyint NOT NULL DEFAULT '0',
+                         PRIMARY KEY (`user_id`),
+                         UNIQUE KEY `login_UNIQUE` (`login`)
+) ENGINE=InnoDB AUTO_INCREMENT=79 DEFAULT CHARSET=utf8;
```
Comment (dmytro-hryhoruk) [inline src/main/resources/init_db.sql]: let's make all create statements look the same and check the order. For example. You cannot create table user_roles before tables users and roles. I should be able to copy paste this file into workbench and recreate your schema

Comment (vremere-a) [reply src/main/resources/init_db.sql]: sorry. my bad
fixed

File: pom.xml [hunk 5]
```diff
@@ -39,6 +39,11 @@
             <artifactId>javax.servlet-api</artifactId>
             <version>3.1.0</version>
         </dependency>
+        <dependency>
+            <groupId>log4j</groupId>
+            <artifactId>log4j</artifactId>
+            <version>1.2.17</version>
+        </dependency>
```
Comment (dmytro-hryhoruk) [inline pom.xml:46]: do you need this dependency? cause I don't see any configurations for logger

File: src/main/java/com/internet/shop/model/User.java (hunk 1 above)
Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/model/User.java:33]: what about other constructors? The screenshot is just an example, let's see where we can use one constructors inside others

Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/model/User.java:33]: And you don't have to add these two constructors if you don't use them

File: src/main/resources/init_db.sql [hunk 6]
```diff
@@ -1,97 +1,122 @@
...
+    `cart_id` BIGINT(11) NOT NULL,
+    `user_id` BIGINT(11) NOT NULL,
+    PRIMARY KEY (`cart_id`),
+    INDEX `carts_users_fk_idx` (`user_id` ASC) VISIBLE,
+    CONSTRAINT `users_carts_fk`
+        FOREIGN KEY (`user_id`)
+            REFERENCES `internet_shop`.`users` (`user_id`)
+            ON DELETE NO ACTION
+            ON UPDATE NO ACTION
+);
```
Comment (dmytro-hryhoruk) [inline src/main/resources/init_db.sql]: first you are creating shopping carts which has foreign key user_id, but table users is not created yet. Check carefully the order of creating tables

File: src/main/java/com/internet/shop/model/User.java (hunk 1 above)
Comment (vremere-a) [reply src/main/java/com/internet/shop/model/User.java:33]: ok.

File: pom.xml (hunk 5 above)
Comment (vremere-a) [reply pom.xml:46]: yes. I used in HashUtil class

File: src/main/resources/init_db.sql (hunk 6 above)
Comment (vremere-a) [reply src/main/resources/init_db.sql]: ok.

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        StringBuilder hashPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(HASH_ALGORITHM);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b : digest) {
+                hashPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            logger.error(e);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/util/HashUtil.java:29]: https://mate-academy.github.io/jv-program-fulltime/02_jdbc_and_servlets/checklist/hw_25.html
and I still can't see any configuration for loggin, so either add it or remove logger with dependency.

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**